numpy/core/src/umath/legacy_array_method.c

```diff
@@ -234,6 +237,94 @@ get_wrapped_legacy_ufunc_loop(PyArrayMethod_Context *context,
+ * trivial for all numeric types.  (Wrapped ufuncs never use byte-swapping.)
+ */
+static int
+copy_cached_initial(
+        PyArrayMethod_Context *context, npy_bool NPY_UNUSED(reduction_is_empty),
+        char *initial)
+{
+    memcpy(initial, context->method->initial, context->descriptors[0]->elsize);
+    return 1;
+}
+
+
+/*
+ * The default `get_reduction_initial` attempts to look up the identity
+ * from the calling ufunc.  This might fail, so we only call it when necessary.
+ *
+ * For our numbers, we can easily cache it, so do so after the first call
+ * by overriding the function with `copy_cache_initial`.  This path is not
+ * publically available.  That could be added, and for a custom initial getter
+ * it will usually be static/compile time data anyway.
+ */
+static int
+get_initial_from_ufunc(
+        PyArrayMethod_Context *context, npy_bool reduction_is_empty,
+        char *initial)
+{
+    if (context->caller == NULL
+            || !PyObject_TypeCheck(context->caller, &PyUFunc_Type)) {
+        /* Impossible in NumPy 1.24;  guard in case it becomes possible. */
+        PyErr_SetString(PyExc_ValueError,
+                "getting initial failed because it can only done for legacy "
+                "ufunc loops when the ufunc is provided.");
+        return -1;
+    }
+    npy_bool reorderable;
+    PyObject *identity_obj = PyUFunc_GetDefaultIdentity(
+            (PyUFuncObject *)context->caller, &reorderable);
+    if (identity_obj == NULL) {
+        return -1;
+    }
+    if (identity_obj == Py_None) {
+        /* UFunc has no idenity (should not happen) */
+        Py_DECREF(identity_obj);
+        return 0;
+    }
+    if (PyTypeNum_ISUNSIGNED(context->descriptors[1]->type_num)
+            && PyLong_CheckExact(identity_obj)) {
+        /*
+         * This is a bit of a hack until we have truly loop specific
+         * identities.  Python -1 cannot be cast to unsigned so convert
+         * it to a NumPy scalar, but we use -1 for bitwise functions to
+         * signal all 1s.
+         * (A builtin identity would not overflow here, although we may
+         * unnecessary convert 0 and 1.)
+         */
+        Py_SETREF(identity_obj, PyObject_CallFunctionObjArgs(
+                     (PyObject *)&PyLongArrType_Type, identity_obj, NULL));
+        if (identity_obj == NULL) {
+            return -1;
+        }
+    }
+    else if (context->descriptors[0]->type_num == NPY_OBJECT
+            && !reduction_is_empty) {
+        /* Allows `sum([object()])` to work, but use 0 when empty. */
+        Py_DECREF(identity_obj);
+        return 0;
+    }
+
+    int res = PyArray_Pack(context->descriptors[0], initial, identity_obj);
+    Py_DECREF(identity_obj);
+    if (res < 0) {
+        return -1;
+    }
+
+    if (PyTypeNum_ISNUMBER(context->descriptors[0]->type_num)) {
+        /* For numbers we can cache to avoid going via Python ints */
+        memcpy(context->method->initial, initial, context->descriptors[0]->elsize);
```

Comment:
So, my sense still would be to just allocate the right number of bytes here, copy over initial to those, and store the pointer (maybe `context->method->cached_initial`?).

Alternatively, making it all even less opaque, I guess one could repack `initial` here as a array or numpy scalar (which now has the right `dtype`) and cache that? Then, the replacement function could simply return the pointer to the data contained in that cached array and in the deallocation, all one would need is a `Py_XDECREF(meth->cached_initial)`.

Reply:
Hmmm, my sense is still that this is simpler as long as it is just a small, private, optimization for now.
To me allocating it feels like we should provide infrastructure so that users can make use of that machinery.  And I am not really motivated to aim for such a mechanism at this point (seems like little gain for additional complexity).

Reply:
If it was hidden in an obviously private place, not in a `struct` that feels like something that should become public, I'd agree!

Reply:
Maybe we need a third opinion :).  Yes we may make (part of) that struct public, and then we would have to hide this away.
But I like keeping the cache and I don't like adding the complexity of an allocation at this time...

Reply:
Sure, this shouldn't hold up what is a really nice PR -- maybe if someone else can have a quick look and see whether it is worth the effort doing an allocation, that would indeed be best. In the end it is really a "does this pass the smell test" matter and my sense of smell is not necessarily the best, especially for C.